# Which Apps Attract the Most Users?

Seeing as our revenue revolves around in-app ads, it seem highly important to see which type of apps attract the most users. Finding out which apps create the most views and engagement within the in-apps ads might entail a story in which we can apply to the future development of apps within the company. 

Therefore, the goal of this analysis is to analyse data that helps the developers understand what type of apps are likely to attract more users and thus create actionable steps to follow or thought provoting points to discuss in the development process (i.e. creating the apps). The apps that are created by the company are for the Google Play and IOS App Store. 

# Exploring the Data

This section aims to find key information on the data that will be useful in the analysis. 

In [1]:
open_file1 = open('AppleStore.csv')
from csv import reader
read_file = reader(open_file1)
apple_data = list(read_file)
apple_data_header = apple_data[0]
apple_data = apple_data[1:]

open_file2 = open('googleplaystore.csv')
from csv import reader
read_file2 = reader(open_file2)
google_data = list(read_file2)
google_data_header = google_data[0]
google_data = google_data[1:]

In the cell above, I opened two data sets; `'AppleStore.csv'` and `'googleplay.csv'`. 

I imported these two data sets into list of lists through the `list()` function assigning them to a more identifiable variable. 

I finally, split the two data sets (which are now list of lists) by excluding the header through variable assignment. 

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
print(google_data_header)
print('\n')
explore_data(google_data, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


In the code above, I created a function named `explore_data()`, which takes four parameters: a dataset; start and end to effectively slice the data; finally a `boolean` which has a default arguement as False. This function enables me to better explore the data sets. 

The number of rows the Google Play data has is: **10841** and the number of columns is: **13**. 

The columns that could possibly help us with our analysis from a quick glance are: **App, Category, Rating, Reviews, Type, Price, and Genre's**. 

More information on the Google Play Store Data Column's can be found [here](https://www.kaggle.com/lava18/google-play-store-apps)

In [3]:
print(apple_data_header)
print('\n')
explore_data(apple_data, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


In the code above, I used the `explore_data()` function to explore the first few rows of the `Apple_data`, count the rows and columns, and `print()` the header. This was was done for the `google_data` in the one above code line. This enabled me to see how many rows and columns there are and see which columns would be best to analysis. 

The number of rows the Apple IOS App Store has is: **7197** and the number of columns is: **16**. 

The columns that could possibly help us with our analysis from a quick glance are: **track_name, price, user_rating, and prime_genre**.  

A better description of each column in the Apple Store Data can be found [here](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

# **Data Cleaning**

The section below is on identifying errors in the Google Play Data. i.e. row errors, duplicate entries...

## Row Error in the Google Play Data:

In [4]:
print(google_data_header)
print('\n')
print(google_data[10472])
print('\n')
print(google_data[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


After reading the discussion section on the Google Play data, I saw that there was a mention of a error in one of the rows. In the above code, I confirmed that row `[10472]` was missing the **category** section. This has an impact on the other columns.

I did this by using the print function to show the header (so I can look at the column names), the row with an error, and a correct row for comparison. I used the `'\n'` function to create a space between each print function int he code.

In [5]:
del google_data[10472]

As this row is an error and can cause problem when analysing the data. It needs to be deleted. This is done in the above code using the `del` function. *It is important that this function is only done once otherwise it will delete multiple rows*

In [6]:
len(google_data)

10840

We can confirm that the `del` function has worked by recounting the number of rows using the `len` function. From our previous count in the code line [2] the number of rows was 10841 but after using the `del` function the number of rows is now 10840. 

The errors was found in the dicussion section, which can be found [here](https://www.kaggle.com/lava18/google-play-store-apps/discussion)

## Duplicate Entries in the Google Play Data:

This section shows that the Google Play Data has duplicate entries. 

**Part One**

In [7]:
for app in google_data:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


As you can see in the above code, that instagram has duplicate entries in the data (four to be exact). 

In [8]:
duplicate_apps = []
unique_apps = []

for app in google_data:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print(len(duplicate_apps))
print('\n')
print(duplicate_apps[:20])

1181


['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express']


In the code above, I created two lists: one for storing the name of duplicate apps, and the other for storing the name of unique apps. Then I looped through the `google_data`, and for each iteration, saved the app name to a variable named `name`. If name was already in the `unique_apps` list, we appended name to the `duplicate_apps` list. Else (if name wasn;t already in the `unique_apps` list), we appended name to the `unique_apps` list. 

We do not want to have duplicate entries, therefore we need to remove them. This could be done randomly, but would likely be highly time consuming and could cause a human error. 

If you were to take a closer look at the instgram duplicates that I listed above. You will see that they vary in number of reviews. Therefore, we could create a criterion on keeping the entry with the highest number of reviews. The reason for the different number of reviews is done to the data being collected at different times. 

We won't remove rows randomly, but rather we'll keep the rows that have the highest number of reviews because the higher the number of reviews, the more reliable the ratings.

In [9]:
print(len(google_data)- 1181)

9659


From looping through the google play data set, we found that there are 1181 duplicates. Using the above calculation, there should be 9659 rows left after the duplicates have been removed. 

**Part two** - Removing the duplicate Entries

In [10]:
reviews_max = {}

for app in google_data: 
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name is not reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))
    
    

9659


The code above, is the first step in removing the duplicate entries. 

The first step, was to create dictionary where each key is a unqiue app anem and the correspinding dictionary value is the highest number of reviews. 
- I made a empty dictionary named `reviews_max`
- Then I looped through the Google Play data set, for each iteration:
- I assgined the app name to a variable called `name` 
- Converted the number of reviews to a `float` as it is currently stored as a string
- If the the `name` already existed as a key in the `reviews_max` dictionary that was created **and** `reviews_max[name] < n_reviews`, I then updated the number of reviews for that entry in the `reviews_max` dictionary
- If `name` **is not in** the `reviews_max` dictionary as a key, I created a new entry in the dictionary where the key is the app name, and the value is the number of reviews. I made sure not to use a `else` clause here, otherwise the number of reviews will be incorrectly updated whenever `reviews_max[name] < n_reviews` evaluates to `False`

I then inspected the dictionary, to ensure it had the correct amount of rows left. 

In [11]:
google_clean = []
already_added = []

for app in google_data:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        google_clean.append(app)
        already_added.append(name)
        
explore_data(google_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


The code above, is the next step in cleaning the data of the duplicate entries.

The aim was to use the dictionary that I created (`reviews_max`) to the remove the duplicate rows. This was done by: 
- creating two empty lists: `google_clean` (which stores our new cleaned data set) and `already_added` (which just stores app names) 
- Then I looped through the Google Play data set, and for each iteration: Assigned the app name to a vairable named `name`; converted the number of reviews to a `float`, and assigned it to a variable named `n_reviews` 
- Then if `n_reveiws` was the same as the number of maximum reviews of the app `name` **and** `name` is not already in the list `already_added`: it appended the entire row to the `google_clean` list (which is a list of list and stores our cleaned data set); and appends the name of the app `name` to the `already_added` list (this helps us keep track of the apps that have already been added). 

Then I used the explore function that was created at the beginning to ensure everything went as expected. The data set has **9,659** rows. 

In [12]:
duplicate_apps1 = []
unique_apps1 = []

for app in apple_data:
    name = app[0]
    if name in unique_apps1:
        duplicate_apps1.append(name)
    else:
        unique_apps1.append(name)

print(len(duplicate_apps1))

0


**The code above, checks for duplicates in the apple data. From the results above, you can see that there are no duplicates for the apple data set for IOS apps.**

# Removing Non-Engligh Apps

## Part 1

From exploring the data - I have found that there is many apps that do not have englsh names. I want to analyze the data of apps that are directed toward english-speakiing audiences. Therefore, I am not interested in keeping these apps and will remove them. 

Both data sets have non-enlgish app names. 

THe way I will be going about this is to remove each app with a name containing a sumbol that is not commmonly used in English text - as english text usially includes letters from the english alphabet, numbers composed of digits from 0 - 9, punctuation marks (., !, ?, ;, etc.), and other symbols. 

Behind the scenes in Python, each character we use in a string has a corresponding number associated with it. For example: `a` is 97. You can get the number of each character using the `ord()` built in function. 

The numbers corresponding to the characters we commonly use in an English text are all in the range 0 - 127. Using this, then we can build a function that detects whether a character belongs to the set of common english characters or not. 

In [13]:
def is_english(string):
    
    for character in string:
        if ord(character) > 127:
            return False
    
    return True

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


In the code above, I created a function that takes in a string and returns `False` if there's any character in the string that doesn't belong to the set of common English characters, otherwise it returns `True`. 

Inside the function, I iterated over the input string. Therefore, for each iteration, it checks whether the number associated with the character is greater than 127. When a character is greater than 127, the function will `return Flase` - The app name is probably not Enlgish, since it contains characters that do not belong to the English alphabet or commonly used symbols. 

If the loop finishes running without the `return` statement being executed, then it means no character had a corresponding number over 127 - the app name is more than likely to be English, the function returns `True`. 

I then checked that the function worked in the last four lines of code. Which it did but with a slight fault. 

The function is not able to detect the use of emojis or other symbols. This would cause a large amount of data being lost. 

Therefore, below we will adapt the coded function. 

In [14]:
def is_english(string):
    non_english = 0
    
    for character in string:
        if ord(character) > 127:
            non_english += 1
    
    if non_english > 3:
        return False
    else:
        return True

print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
True


The function is still not perfect, and very few non-English apps might get past our filter, but this seems good enough at this point in my analysis — I shouldn't spend too much time on optimization at this point.

The function was changed so if the input string has more than 3 characters that fall outside the range (0- 127), then the function should return `False`, otherwise it should return `True`. 

Using the print function, an example was used to ensure it worked.

In [15]:
english_google = []

for apps in google_clean:
    name = apps[0]
    if is_english(name):
        english_google.append(apps)

explore_data(english_google, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


In [16]:
english_apple = []

for apps in apple_data:
    name = apps[1]
    if is_english(name):
        english_apple.append(apps)

explore_data(english_apple, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 6183
Number of columns: 16


In the above code, we used our new function `is_english()` to filter out the non english apps from both data sets. 

We did this creating a new seperate list, then looping through each set if a app name is identified as english it is then appended to the new list. 

Finally, I explored the data sets, to see how many remaining rows there are using the `explore()` function. 

**So far in the data cleaning process, we have: Removed inaccurate data; Removed duplicate app enties; and Removed non-english apps.**

In [17]:
final_google = []
final_apple = []

for apps in english_google:
    price = apps[7]
    if price == '0':
        final_google.append(apps)
        
for apps in english_apple:
    price = apps[4]
    if price == '0.0':
        final_apple.append(apps)
        
print(len(final_google))
print(len(final_apple))

8864
3222


It is important to remember, that I only want to analyse the 'free apps'. The data set, had both free and paid apps; in the code above, I isolated only the free apps for the analysis. 

This was done very simply, by creating a seperate list, looping through the data sets and then appending the free apps to the newly created lists. 

I then using the `len()` function to see how rows are remaining. 

# Analysing Common Genres

As mentioned above, our aim to find the kinds of apps that are likely to attract more users as the companies revenue is based on how many people use our apps. 

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

The end goal is to have apps in both App markets ( Google Play and IOS APP Store ), therefore we need to find app profiles that are successful in both markets. 

We will begin by getting a sense of what are the most common genres for each market. This will be done by building frequency tables for a few columns in the data sets. 

## Analysis of Genres

In the code below, is a built frequency table for the `prime_genre` column of the app store data set, and for the `Genres` and `Category` columns of the Google Play Data Set. 

The two built functions are used to analyze the frequency tables:

- One function to generate frequency tables that show percentages 
- Another function we can use to display the percentages in a descending order

In [18]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [19]:
display_table(final_apple, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

Let's continue by examining the Genres and Category columns of the Google Play data set (two columns which seem to be related).

In [21]:
display_table(final_google, 1) # Catergory

FAMILY : 19.223826714801444
GAME : 9.510379061371841
TOOLS : 8.461191335740072
BUSINESS : 4.580324909747293
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.5424187725631766
SPORTS : 3.4183212996389893
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2490974729241873
HEALTH_AND_FITNESS : 3.068592057761733
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.782490974729242
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.128158844765343
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
ENTERTAINMENT : 0.8799638989169676
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 0

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column:

In [23]:
display_table(final_google, -4) # Genres

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.580324909747293
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.5424187725631766
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2490974729241873
Action : 3.1024368231046933
Health & Fitness : 3.068592057761733
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.861462093862816
Video Players & Editors : 1.782490974729242
Casual : 1.7486462093862816
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.925090252707581

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

## Most Popular Apps by Genre on the App Store

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

Below, we calculate the average number of user ratings per app genre on the App Store:

In [27]:
genres_ios = freq_table(final_apple, -5)

for genre in genres_ios:
    total = 0 
    len_genre = 0 
    for app in final_apple:
        genre_app = app[-5]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Entertainment : 14029.830708661417
Shopping : 26919.690476190477
Education : 7003.983050847458
Book : 39758.5
Lifestyle : 16485.764705882353
Reference : 74942.11111111111
Productivity : 21028.410714285714
Weather : 52279.892857142855
Sports : 23008.898550724636
News : 21248.023255813954
Photo & Video : 28441.54375
Medical : 612.0
Catalogs : 4004.0
Navigation : 86090.33333333333
Social Networking : 71548.34905660378
Finance : 31467.944444444445
Utilities : 18684.456790123455
Health & Fitness : 23298.015384615384
Music : 57326.530303030304
Games : 22788.6696905016
Travel : 28243.8
Food & Drink : 33333.92307692308
Business : 7491.117647058823


On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together.

The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating. 

However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.

## Most Popular Apps by Genre on Google Play
 
For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [29]:
display_table(final_google, 5) # the Installs columns

1,000,000+ : 15.749097472924186
100,000+ : 11.563628158844766
10,000,000+ : 10.503158844765343
10,000+ : 10.209837545126353
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


On average, communication apps have the most installs: 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs.


If we removed all the communication apps that have over 100 million installs, the average would be reduced roughly ten times.


We see the same pattern for the video players category, which is the runner-up with 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previously we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

The books and reference genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the App Store, and our aim is to recommend an app genre that shows potential for being profitable on both the App Store and Google Play.

Let's take a look at some of the apps from this genre and their number of installs. 


The book and reference genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average. 

However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads). 


This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

## Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.